esp/services/ws_workunits/ws_wudetails.cpp

```diff
@@ -0,0 +1,349 @@
+          if ( *sStatisticKind==0 ) continue; // skip nulls
+
+          StatisticKind statisticKind = queryStatisticKind(sStatisticKind);
+          if (statisticKind == StKindNone || statisticKind== StKindAll)
+              throw MakeStringException(ECLWATCH_INVALID_INPUT,
+                                        "Invalid Attribute Name ('%s') in Attribute Filter",
+                                        reqAttribFilter.item(idx).getName());
+
+          const char * sExactValue = attr.getExactValue();
+          const char * sMinValue = attr.getMinValue();
+          const char * sMaxValue = attr.getMaxValue();
+          bool bExactValue = *sExactValue!=0;
+          bool bMinValue = *sMinValue!=0;
+          bool bMaxValue = *sMaxValue!=0;
+
+          if (bExactValue && (bMinValue||bMaxValue))
+              throw MakeStringException(ECLWATCH_INVALID_INPUT,
+                                        "Invalid Attribute Filter ('%s') - ExactValue may not be used with MinValue or MaxValue",
+                                        reqAttribFilter.item(idx).getName());
+
+          Owned<CAttributeFilter> caf = new CAttributeFilter(statisticKind, atoi64(sExactValue), atoi64(sMinValue),
+                                                             atoi64(sMaxValue), bExactValue, bMinValue, bMaxValue);
+          processedAttribFilter.append(*(caf.getLink()));
+    }
+    processedAttribFilter.sort(funcCompareAttribFilter);
+}
+
+void CWUDetails::buildAttribListToReturn(UnsignedShortArray & skReturnAttribList,
+                                         StringArray & returnAttriblist)
+{
+    bool isNew;
+    ForEachItemIn(idx,returnAttriblist)
+    {
+        const char * attribName = returnAttriblist[idx];
+        if (*attribName==0) continue;
+
+        StatisticKind sk = queryStatisticKind(attribName);
+        if (sk==StKindNone || sk==StKindAll)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid Attribute name in AttributeToReturn(%s)",attribName);
+
+        unsigned short usk = static_cast<unsigned short> (sk);
+        skReturnAttribList.bAdd(usk, funcCompareUnsignedShort, isNew);
+    }
+
+}
+
+void CWUDetails::buildScopeTypeList(UnsignedShortArray & processedScopeTypeList,
+                                    StringArray & arrayScopeTypeList)
+{
+    bool isNew;
+    ForEachItemIn(idx, arrayScopeTypeList)
+    {
+        const char * scopeType = arrayScopeTypeList[idx];
+        if (*scopeType==0) continue;
+
+        StatisticScopeType sst = queryScopeType(scopeType);
+        if (sst==SSTnone ||sst==SSTall)
+            throw MakeStringException(ECLWATCH_INVALID_INPUT, "Invalid ScopeType(%s) in filterNestedScopeTypes",scopeType);
+
+        unsigned short usst = static_cast<unsigned short> (sst);
+        processedScopeTypeList.bAdd(usst, funcCompareUnsignedShort, isNew);
+    }
+}
+
+void CWUDetails::processWUStatsIntoResponse(IArrayOf<IEspWUScope> & respScopes,
```

Comment:
It depends on what your class is meant to model.  If it is a single instance of the service then they are probably cleaner as member variables.  Returning multiple values from a function via references is always a potential for confusion.